### PR TITLE
Fix sync_dist epoch metrics inflated by floored integer batch-size accumulator

### LINKED ISSUE #21859: sync_dist mean-reduced epoch metrics are inflated when ranks log unequal batch counts (integer batch-size accumulator floors during mean-sync)
### Bug description

Epoch-level metrics logged with `self.log(..., on_epoch=True, sync_dist=True)` are **uniformly inflated** whenever ranks accumulate unequal cumulative batch sizes — e.g. an `IterableDataset` whose shards don't divide evenly across ranks, or uneven last batches. Every affected metric is scaled by the same factor `true_denominator / floor(true_denominator / world_size * world_size ... )` — concretely: `(Σbs / W) / floor(Σbs / W)` where `Σbs` is the total logged batch-size mass and `W` the world size.

**Root cause.** `_ResultMetric` accumulates `value` (float) and `cumulated_batch_size` (**int64**, created as `torch.tensor(0)`), and `compute()` mean-syncs *both* through the strategy's `reduce`:

https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L246-L253

A mean-reduce over an **integer** tensor floors:

- **NCCL**: `_sync_ddp` maps `"mean"` → `ReduceOp.AVG`, and NCCL's AVG on integral dtypes is sum-then-**integer**-division.
- **gloo**: `_sync_ddp` does SUM + `result.copy_(result / world_size)` — the true division result is copied back **into the integer tensor**, truncating:

https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/utilities/distributed.py#L218-L223

The float numerator keeps its exact fractional value while the integer denominator floors, so the division no longer cancels and the returned mean is inflated. When `Σbs` is divisible by `W` the floor is lossless and everything looks correct — which is why this survives most 2/4/8-rank setups and unit tests, and then silently corrupts larger world sizes.

**Real-world impact (how we found it):** on a 16-rank (4 nodes × 4 GPU) run with a 44-sample validation split logged with `batch_size=1`, every `val` metric was logged ×(44/16)/floor(44/16) = **1.375× too high**, while the identical code on 4 ranks (44/4 = 11, divisible) reproduced single-process values to ≤1e-4. Because *all* mean-reduced metrics scale by the same factor, "higher-is-better" metrics look too good and "lower-is-better" metrics look too bad simultaneously, which is quite misleading during analysis — and any cross-run comparison or val-keyed automation (checkpoint selection across world sizes, early stopping thresholds) silently operates on wrong values.

### Proposed fix

Sync `cumulated_batch_size` in the value's floating dtype so the denominator keeps its fractional part:

```python
cumulated_batch_size = self.meta.sync(self.cumulated_batch_size.to(value.dtype))
```

This keeps the accumulator state itself integer (no checkpoint/state-dict implications) and fixes the reproduction above (returns 7.2). I have a PR ready with this fix plus a 2-rank regression test in `tests_pytorch/core/test_results.py`.

**Related observation** (left out of the PR to keep it scoped): `_sync_ddp` itself floors integer tensors for `"mean"`/`"avg"` on both backends, and `_test_all_reduce` in `tests_fabric/utilities/test_distributed.py` currently *encodes* the floored expectation for integer dtypes (the expected value is cast to the integer dtype before comparison). If maintainers consider integer-mean flooring in `strategy.reduce` itself a bug rather than a contract, I'm happy to follow up separately — it changes the documented in-place semantics (`result is tensor`), so it deserves its own discussion.

### What version are you seeing the problem on?

v2.6, master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader, IterableDataset

import lightning.pytorch as pl


class RankSizedDataset(IterableDataset):
    """Rank 0 yields 3 samples, rank 1 yields 2 (e.g. sharded data that doesn't divide evenly)."""

    def __iter__(self):
        rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
        values = ([1.0, 2.0, 3.0], [10.0, 20.0])[rank]
        return iter([torch.tensor([v]) for v in values])


class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(1, 1)

    def forward(self, x):
        return self.layer(x)

    def validation_step(self, batch, batch_idx):
        self.log("val_metric", batch.mean(), on_epoch=True, sync_dist=True, batch_size=1)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.1)


if __name__ == "__main__":
    trainer = pl.Trainer(accelerator="cpu", devices=2, strategy="ddp_spawn", logger=False,
                         enable_progress_bar=False, enable_model_summary=False)
    results = trainer.validate(Model(), DataLoader(RankSizedDataset(), batch_size=None))
    expected = (1 + 2 + 3 + 10 + 20) / 5  # 7.2, the batch-size-weighted mean over all logged batches
    print(f"expected {expected}, got {results[0]['val_metric']}")
```

### Error messages and logs

```
expected 7.2, got 9.0
```

### Environment

<details>
  <summary>Current environment</summary>

```
- Lightning: master (35c8970ba) and 2.6.0
- PyTorch: 2.9.1
- Verified on CPU/gloo (script above); NCCL path hit in production on 16 ranks (4 nodes × 4× A10G/L4)
```

</details>

### More info

[Fix PR](https://github.com/Lightning-AI/pytorch-lightning/pull/21839)

cc @ethanwharris @lantiga

### PR DESCRIPTION
**Pull request by nmirman**
## What does this PR do?

Fixes #21859

`_ResultMetric.compute()` mean-syncs both the metric `value` (float) and the `cumulated_batch_size` accumulator (**int64**) through `strategy.reduce`. A mean-reduce over an integer tensor floors — NCCL's `ReduceOp.AVG` divides integral dtypes with integer arithmetic, and the gloo fallback copies the divided result back into the integer tensor — so whenever ranks accumulate unequal cumulative batch sizes, the floored denominator uniformly inflates every `self.log(..., on_epoch=True, sync_dist=True)` metric. When the total batch-size mass divides evenly by the world size the floor is lossless, which is why this goes unnoticed on typical small world sizes and then silently corrupts larger ones (we hit it as a 1.375× inflation of every validation metric on a 16-rank run whose 44-sample val split didn't divide evenly).

The fix syncs the accumulator in the value's floating dtype so the denominator keeps its fractional part:

```python
cumulated_batch_size = self.meta.sync(self.cumulated_batch_size.to(value.dtype))
```

The stored accumulator state remains integer (no state-dict/checkpoint implications; existing tests asserting the raw accumulator state are unaffected). Includes a 2-rank regression test in `tests_pytorch/core/test_results.py` that fails with 9.0 (floored) before the fix and returns the exact weighted mean 7.2 after.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs) → yes, #21838
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary) → not necessary
- Did you write any **new necessary tests**? (not for typos and docs) → yes
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request? → none
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors) → yes

</details>

## PR review

Anyone in the community is welcome to review the PR.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/lightning/pytorch/CHANGELOG.md [hunk 1]
```diff
@@ -32,6 +32,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed mean-reduced `self.log(..., sync_dist=True, on_epoch=True)` metrics being uniformly inflated when ranks accumulate unequal cumulative batch sizes (e.g. uneven batch counts across an `IterableDataset`), caused by the integer batch-size accumulator being floored during the cross-rank mean-sync ([#21838](https://github.com/Lightning-AI/pytorch-lightning/issues/21838))
```
Comment (deependujha) [inline src/lightning/pytorch/CHANGELOG.md]:
```suggestion
- Fixed mean-reduced `self.log(..., sync_dist=True, on_epoch=True)` metrics being uniformly inflated when ranks accumulate unequal cumulative batch sizes (e.g. uneven batch counts across an `IterableDataset`), caused by the integer batch-size accumulator being floored during the cross-rank mean-sync ([#21839](https://github.com/Lightning-AI/pytorch-lightning/issues/21839))
```

File: tests/tests_pytorch/core/test_results.py [hunk 2]
```diff
@@ -54,3 +54,24 @@ def result_reduce_ddp_fn(strategy):
...
+    metadata = _Metadata("validation_step", "val_metric", on_step=False, on_epoch=True)
+    metadata.sync = _Sync(strategy.reduce, _should=True)
+    metric = _ResultMetric(metadata, is_tensor=True)
+    values = ([1.0, 2.0, 3.0], [10.0, 20.0])[dist.get_rank()]
+    for value in values:
+        metric.update(torch.tensor(value), batch_size=1)
+    assert metric.compute().item() == pytest.approx(7.2)
+
+
+@pytest.mark.flaky(reruns=3)
```
Comment (deependujha) [inline tests/tests_pytorch/core/test_results.py:73]: why is it marked flaky?

Comment (deependujha) [reply tests/tests_pytorch/core/test_results.py:73]: got it, it's been a pattern based on flakiness. I believe using `standalone=True` should work and will be better alternative than marking them as flaky.

It's fine for this PR, but a quick discussion and verification would be helpful. Just a thought. cc: @bhimrazy

**Review by deependujha — APPROVED** [review]

File: tests/tests_pytorch/core/test_results.py (hunk 2 above)
Comment (nmirman) [reply tests/tests_pytorch/core/test_results.py:73]: Hi @deependujha , for context it was marked flaky for the spawn/gloo transport only, to mitigate transient init failures in CI.  The assertion itself is deterministic.  No strong opinion on flaky vs `standalone=True`, with the tradeoff being that the standalone suite runs in fewer CI jobs.  Other than that I don't see an issue with changing to `standalone=True` if that's preferable.

**Review by bhimrazy — APPROVED** [review]

**Comment by nmirman** [comment]
Bumping this -- is there anything else we should do to make it ready to merge?
@ethanwharris @tchaton @justusschock

**Comment by chrismbryant** [comment]
Thanks for the PR! Do we have an estimate of when this can get merged?
